Upgrade to jline3

eatkins opened the pull request:
This PR upgrades sbt to jline 3 primarily so that we can get better tab completions in the sbt shell:
[![asciicast](https://asciinema.org/a/2uQZpOSgLke0RYvKMfPF1coEC.svg)](https://asciinema.org/a/2uQZpOSgLke0RYvKMfPF1coEC)

We still need jline2 for the console task to work with most scala versions. This made the PR a bit bigger in terms of net lines than I would have liked because I had to leave in all of the old jline2 related stuff. JLine3 also has some strange quirks around creating terminals and the input streams that it consumed that required some unusual workarounds.

In addition to the improved tab completions. I also fixed a few other small ui issues. Now when a client connects to a loading server, they will see all of the log messages since the server started which is especially useful if the server failed to load due to a misconfigured build. I also improved the supershell logging so that it works more reliably across the full range of scenarios that I test.

eatkins committed:
Upgrade LineReader to JLine3

This commit upgrades sbt to using jline3. The advantage to jline3 is
that it has a significantly better tab completion engine that is more
similar to what you get from zsh or fish.

The diff is bigger than I'd hoped because there are a number of
behaviors that are different in jline3 vs jline2 in how the library
consumes input streams and implements various features. I also was
unable to remove jline2 because we need it for older versions of the
scala console to work correctly with the thin client. As a result, the
changes are largely additive.

A good amount of this commit was in adding more protocol so that the
remote client can forward its jline3 terminal information to the server.

There were a number of minor changes that I made that either fixed
outstanding ui bugs from #5620 or regressions due to differences between
jline3 and jline2.

The number one thing that caused problems is that the jline3 LineReader
insists on using a NonBlockingInputStream. The implementation ofo
NonBlockingInputStream seems buggy. Moreover, sbt internally uses a
non blocking input stream for system in so jline is adding non blocking
to an already non blocking stream, which is frustrating.

A long term solution might be to consider insourcing LineReader.java
from jline3 and just adapting it to use an sbt terminal rather than
fighting with the jline3 api. This would also have the advantage of not
conflicting with other versions of jline3. Even if we don't, we may want to
shade jline3 if that is possible.

eatkins committed:
Fix copied scaladoc

eatkins committed:
Add loop to SelectMainClass

Rather than exiting when the user enters an invalid line, we can
reprompt them.

eatkins committed:
Buffer terminal lines

It is useful to store a buffer of the lines written to each terminal. We
can use those lines to replay the terminal log lines to a different
client. This is particularly nice when a remote client connects to sbt
while it's booting. We can show the remote client all the lines
displayed by the console prior to the client connecting.

eatkins committed:
Return correct path in native executable task

Bonus: throw an error if native building fails

eatkins committed:
Move ProgressState into its own file

This didn't really belong in ConsoleAppender.scala anymore.

eed3si9n commented:
Nice! Thanks for the contribution @eatkins.

eed3si9n commented on internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala:
```diff
@@ -8,24 +8,98 @@
+  ): LineReader = {
+    val term = JLine3(terminal)
+    // We may want to consider insourcing LineReader.java from jline. We don't otherwise
+    // directly need jline3 for sbt.
```
I'm a bit confused by this statement. LineReader.java (https://github.com/jline/jline3/blob/3b0d5d52fd2c31fb92ad8aaa2f11a22aceab890a/reader/src/main/java/org/jline/reader/LineReader.java) is an empty interface. Don't we need JLine's implementation for the reader?

eatkins replied:
Yeah. I meant the implementation class https://github.com/jline/jline3/blob/3b0d5d52fd2c31fb92ad8aaa2f11a22aceab890a/reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java. The idea would just be to make a partial fork of jline3 that only provides the minimal subset of functionality from `LineReaderImpl` that we need to make line reading and tab completion work but without using all of other jline 3 stuff that we neither want or need. I'm not sure how excited I am about actually doing that though.

eed3si9n replied:
What's the diff for `show sbtProj/Compile/externalDependencyClasspath` pre and post this PR?

eatkins replied:
```patch
+ Attributed($MAVEN/org/jline/jline/3.15.0/jline-3.15.0.jar)
```

eatkins committed:
Cross build collectionProj

eatkins commented:
~~The console task is slightly broken at the moment. Left arrow doesn't work correctly probably due to a jline2 vs jline3 interaction that I need to track down.~~

Fixed in https://github.com/sbt/sbt/pull/5671/commits/1749e45bb11af1b71f1df52b9844e5186560b06f

eatkins committed:
Unprompt channels during project load

In the situation where sbt was started in server mode and a client is
running a `~` command and a project reload is triggered by a change to
a build source, the console terminal looks like

sbt:foo>
[info] received remote command: ~compile
sbt:foo>
[info] welcome to sbt 1.4.0-SNAPSHOT (Azul Systems, Inc. Java 1.8.0_252)
sbt:foo>
[info] loading global plugins from ~/.sbt/1.0/plugins
sbt:foo>
[info] loading settings for project foo-build from metals.sbt ...
sbt:foo>
[info] loading project definition from
~/foo/project
sbt:foo>
[info] loading settings for project root from build.sbt ...
sbt:foo>
[info] loading settings for project macros from build.sbt ...
sbt:foo>
[info] loading settings for project main from build.sbt ...
sbt:foo>
[info] set current project to foo (in build file:~/foo)
sbt:foo>

This change fixes that by unprompting all channels during project
loading and reprompting them when it completes.

eatkins committed:
Only trigger reload if sources have changed

Running a `~` command in a local build off the latest develop branch
will cause the build to reload even if the build sources were only
touched and not actually modified.

eatkins committed:
Set terminal before watch triggered reload

Without setting the terminal, a remote client is unable to reload the
project if it fails.

eatkins committed:
Clarify boolean flag

I found this difficult to read.

eatkins committed:
Don't log systemOut messages in jsonRpcNotify

Whatever debugging utility these may have is not worth spamming the task
logs with these.

eatkins committed:
Add Terminal.withRawOutput api

In the scala console, it's essential that we not process the bytes that
are written to the terminal by jline.

eatkins committed:
Use terminal printstream in CheckBuildSources

The build source check is evaluated at times when we can't be completely
sure that global logger is pointing at the terminal that initiated the
reload (which may be a passive watch client). To work around this, we
can inspect the exec to determine which terminal initiated the check and
write any output directly to that terminal.

eatkins committed:
Buffer output to the remote client

Remote clients sometimes flicker when updating progress. This is especially
noticeable when there are two clients and one of them is running a command,
the other will tend to have some visible flickering and character ghosting.
As an experiment, I buffered calls to flush in the NetworkChannel output
stream and the artifacts went away.

eatkins committed:
Aggregate watch events

It is possible for sbt to get into a weird state when in a continuous
build when the auto reload feature is on and a source file and a build
file are changed in a small window of time. If sbt detects the source
file first, it will start running the command but then it will
autoreload when it runs the command because of the build file change.
This causes the watch to get into a broken state because it is necessary
to completely restart the watch after sbt exits.

To fix this, we can aggregate the detected events in a 100ms window. The
idea is to handle bursts of file events so we poll in 5ms increments and
as soon as no events are detected, we trigger a build.

eatkins committed:
Don't start an ask user thread for dead console

If there is no system console available, then there is no point in
making an ask user thread. An ask user thread can only be created when
the terminal prompt is in the Prompt.Running or Prompt.Loading state.
The console channel will now set itself to be in the Prompt.NoPrompt
state if it detects that there is no System.console available.

The motivation for this change is that jline was printing a lot of extra
text during scripted and server tests. Whenever a jline3 linereader is
closed, it prints a newline so the logs were filled with unnecessary
newlines.
